infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogDataStore.java

```diff
@@ -1,119 +1,276 @@
+            return ZERO_ADDRESS;
+        });
     }
 
     /**
-     * Returns the dataStore starting address.
+     * Return the current head segment.
      *
      * @return the starting address
      */
-    public long getStartingAddress() {
-        if (startingAddress.get() == Address.NON_ADDRESS) {
-            startingAddress.set(dataStore.get(STARTING_ADDRESS_RECORD, ZERO_ADDRESS));
+    long getHeadSegment() {
+        return headSegment.get();
+    }
+
+    /**
+     * Update current head segment in the data store.
+     *
+     * @param newHeadSegment new head segment to update
+     */
+    void updateHeadSegment(long newHeadSegment) {
+        if (update(HEAD_SEGMENT_RECORD, headSegment, newHeadSegment, (curr, next) -> next < curr)) {
+            log.debug("Updated head segment to: {}", newHeadSegment);
+        } else {
+            log.trace("New head segment {} not smaller than current, ignore.", newHeadSegment);
+        }
+    }
+
+    /**
+     * Reset head segment.
+     */
+    void resetHeadSegment() {
+        log.info("Resetting head segment. Current head segment: {}", headSegment.get());
+        headSegment.updateAndGet(addr -> {
+            dataStore.put(HEAD_SEGMENT_RECORD, Address.MAX);
+            return Address.MAX;
+        });
+    }
+
+    /**
+     * Return the current global compaction mark.
+     */
+    long getGlobalCompactionMark() {
+        return globalCompactionMark.get();
+    }
+
+    /**
+     * Update the current global compaction mark if the provided address is greater.
+     *
+     * @param newCompactionMark a new address to update current global compaction mark
+     */
+    void updateGlobalCompactionMark(long newCompactionMark) {
+        if (update(COMPACTION_MARK_RECORD, globalCompactionMark, newCompactionMark, (curr, next) -> next > curr)) {
+            log.debug("Updated global compaction mark to: {}", newCompactionMark);
+        } else {
+            log.trace("New global compaction mark {} not greater than current, ignore.", newCompactionMark);
         }
+    }
+
+    /**
+     * Reset global compaction mark.
+     */
+    void resetGlobalCompactionMark() {
+        log.info("Resetting global compaction mark. Current: {}", globalCompactionMark.get());
+        globalCompactionMark.updateAndGet(addr -> {
+            dataStore.put(COMPACTION_MARK_RECORD, Address.NON_ADDRESS);
+            return Address.NON_ADDRESS;
+        });
+    }
 
-        return startingAddress.get();
+    /**
+     * Return the current committed log tail.
+     */
+    long getCommittedTail() {
+        return committedTail.get();
     }
 
     /**
-     * Update current starting address in the data store
+     * Update the current committed log tail if the provided tail is greater.
      *
-     * @param newStartingAddress updated starting address
+     * @param newCommittedTail a new address to update current global committed tail
+     */
+    void updateCommittedTail(long newCommittedTail) {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use explicit scoping instead of the default package private level](https://app.codacy.com/app/annymartinez.m/CorfuDB/pullRequest?prid=4651450)